src/NodaTime/Text/CompositePatternBuilder.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright 2017 The Noda Time Authors. All rights reserved.
+// Use of this source code is governed by the Apache License 2.0,
+// as found in the LICENSE.txt file.
+
+using JetBrains.Annotations;
+using NodaTime.Utility;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace NodaTime.Text
+{
+    /// <summary>
+    /// A builder for composite patterns.
+    /// </summary>
+    /// <remarks>
+    /// A composite pattern is a combination of multiple patterns. When parsing, these are checked
+    /// in the order in which they are added to the builder with the <see cref="Add(IPattern{T}, Func{T, bool})"/>
+    /// method, by trying to parse and seeing if the result is a successful one. When formatting,
+    /// the patterns are checked in the reverse order, using the predicate provided along with the pattern
+    /// when calling <c>Add</c>. The intention is that patterns are added in "most precise first" order,
+    /// and the predicate should indicate whether it can fully represent the given value - so the "less precise"
+    /// (and therefore usually shorter) pattern can be used first.
+    /// </remarks>
+    /// <typeparam name="T">The type of value to be parsed or formatted by the resulting pattern.</typeparam>
+    internal sealed class CompositePatternBuilder<T> : IEnumerable<IPattern<T>>
```

Comment:
This was meant to be public... but if we submit while it's internal, we can still reflect on the API a bit further.

I'll add some more tests, fix whitespace-only changes and then merge.